### PR TITLE
azurestackprivate Terraform Provider

### PR DESCRIPTION
**Pull request by patrickdillon**
This PR creates a private terraform provider for uploading RHCOS images to Azure Stack Hub environments as part of the install process. Based on the azurestack terraform provider and the [azure-vhd-utils](https://github.com/microsoft/azure-vhd-utils) project, the majority of the original work here is in the /pkg/terraform/exec/plugins/azurestackprivate/resource_vhd_blob.go, particularly the `resourceArmVHDBlobCreate` function. 

The main piece of feedback needed is regarding updating azure-vhd-utils to import openshift/azure-sdk-for-go to get around the antiquated version dependency. This is described in detail in the commit messages.

Status: This PR stands on its own and can be reviewed/merged but the overall image upload process still needs to be completed
- [x] private terraform provider to upload vhd
- [x] terraform configs to make use of the provider (testing has been completed)
- [x] tfvars work in the installer to cache rhcos image

### CODE REVIEW

**Comment by patrickdillon** [comment]
I will address the linting/vetting errors. 

The main piece of tech debt this would incur is that we would not be able to use our fork of azure-sdk-for-go without resolving these issues. We don't use our fork at the moment, so this is a hypothetical problem.

**Comment by patrickdillon** [comment]
Refactored a little which should hopefully make the go-vet test pass. Still need to fix go-lint. 

Was able to reduce the amount of code we are copying directly into the provider.

**Comment by patrickdillon** [comment]
Depends on #5084 for the TF config

File: pkg/terraform/exec/plugins/azurestackprivate/resource_vhd_blob.go [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+	storageAccountKey, accountExists, err := armClient.getKeyForStorageAccount(ctx, resourceGroupName, storageAccountName)
+	if err != nil {
+		return err
+	}
+	if !accountExists {
+		return fmt.Errorf("storage account %q not found in resource group %q while uploading RHCOS image", storageAccountName, resourceGroupName)
+	}
+
+	blobServiceBaseURL := strings.TrimPrefix(armClient.environment.ResourceManagerEndpoint, "https://management.")
+	storageClient, err := storage.NewClient(storageAccountName, storageAccountKey, blobServiceBaseURL, apiProfileVersion, true)
```
[the remark this answers is not in the document]
Comment (patrickdillon) [reply pkg/terraform/exec/plugins/azurestackprivate/resource_vhd_blob.go:93]: Was wondering if there was a programmatic way to pull that baseURL directly.

Should also make sure that the blobServiceBaseURL will always equal the base url of the ARM endpoint.

**Review by staebler — COMMENTED** [review]
Patrick and I had a discussion about exploring the possibility of embedding a azure-vhd-utils binary that could be run as part of a local-exec resource rather than having to bring into and own the azure-hvd-utils code in the installer repo.

File: data/data/azurestack/vnet/main.tf [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+resource "azurestack_image" "cluster" {
+  name                = var.cluster_id
+  resource_group_name = data.azurestack_resource_group.main.name
+  location            = var.azure_region
+
+  os_disk {
+    os_type  = "Linux"
+    os_state = "Generalized"
+    blob_uri = local.rhcos_blob_uri
```
Comment (freedge) [inline data/data/azurestack/vnet/main.tf:99]: this creates an image from a blobUri which seems not always possible in all versions of AzureStack. I personally need to create a managed disk first, in this fashion (bicep):
```
resource imgdisk 'Microsoft.Compute/disks@2019-03-01' = {
  name: 'fedora-coreos-disk'
  location: location
  sku:{
    name: 'Premium_LRS'
  }
  properties: {
    creationData: {
      createOption: 'Import'
      sourceUri: blobUrl
    }
    diskSizeGB: 150
  }
}

resource img 'Microsoft.Compute/images@2017-12-01' = {
  location: location
  name: 'fedoracos'
  properties: {
    storageProfile: {
      osDisk: {
        osType: 'Linux'
        osState: 'Generalized'
        managedDisk: {
          id: imgdisk.id
        }
        storageAccountType: 'Premium_LRS'
      }
    }
  }
}
```

File: pkg/terraform/exec/plugins/azurestackprivate/resource_vhd_blob.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (patrickdillon) [reply pkg/terraform/exec/plugins/azurestackprivate/resource_vhd_blob.go:93]: That seems correct.

**Comment by patrickdillon** [comment]
/close

We will add a field in the install config for an image URL but we expect users to upload first. Hopefully the burden will be mitigated in the future by making the images available in the Azure marketplace.

**Pull request closed without merging**